libgalois/include/katana/PropertyGraph.h

```diff
@@ -473,80 +487,200 @@ class KATANA_EXPORT PropertyGraph {
+    return edge_entity_type_id < edge_entity_type_id_to_type_names_.size();
+  }
+
+  /// \returns the node EntityTypeID for an atomic node type with name
+  /// @param name
   /// (assumes that the node type exists)
-  const SetOfTypeSetIDs& NodeTypeNameToTypeSetIDs(
-      const std::string& name) const {
-    return node_type_name_to_type_set_ids_.at(name);
+  bool NodeAtomicTypeNameToEntityTypeID(const std::string& name) const {
+    return node_atomic_type_name_to_entity_type_id_.at(name);
   }
 
-  /// \returns the set of edge TypeSetIDs that contain
-  /// the edge type with @param name
+  /// \returns the edge EntityTypeID for an atomic edge type with name
+  /// @param name
   /// (assumes that the edge type exists)
-  const SetOfTypeSetIDs& EdgeTypeNameToTypeSetIDs(
-      const std::string& name) const {
-    return edge_type_name_to_type_set_ids_.at(name);
+  bool EdgeAtomicTypeNameToEntityTypeID(const std::string& name) const {
+    return edge_atomic_type_name_to_entity_type_id_.at(name);
+  }
+
+  /// \returns the name of the atomic type if the node EntityTypeID
+  /// @param node_entity_type_id is an atomic type,
+  /// nullopt otherwise
+  /// (assumes that the node EntityTypeID exists)
+  std::optional<std::string> NodeEntityTypeIDToAtomicTypeName(
+      EntityTypeID node_entity_type_id) const {
+    auto& set_of_names =
+        node_entity_type_id_to_type_names_.at(node_entity_type_id);
+    if (set_of_names.size() == 1) {
+      return *set_of_names.cbegin();
+    }
+    return std::nullopt;
+  }
+
+  /// \returns the name of the atomic type if the edge EntityTypeID
+  /// @param edge_entity_type_id is an atomic type,
+  /// nullopt otherwise
+  /// (assumes that the node EntityTypeID exists)
+  std::optional<std::string> EdgeEntityTypeIDToAtomicTypeName(
+      EntityTypeID edge_entity_type_id) const {
+    auto& set_of_names =
+        edge_entity_type_id_to_type_names_.at(edge_entity_type_id);
+    if (set_of_names.size() == 1) {
+      return *set_of_names.cbegin();
+    }
+    return std::nullopt;
   }
 
-  /// \returns the number of node TypeSetIDs (including kUnknownType)
-  size_t GetNodeTypeSetIDsNum() const {
-    return node_type_set_id_to_type_names_.size();
+  /// \returns the set of node EntityTypeIDs that contain
+  /// the node type with @param name
+  /// (assumes that the node type exists)
+  const SetOfEntityTypeIDs& NodeTypeNameToEntityTypeIDs(
+      const std::string& name) const {
+    return node_type_name_to_entity_type_ids_.at(name);
   }
 
-  /// \returns the number of edge TypeSetIDs (including kUnknownType)
-  size_t GetEdgeTypeSetIDsNum() const {
-    return edge_type_set_id_to_type_names_.size();
+  /// \returns the set of edge EntityTypeIDs that contain
+  /// the edge type with @param name
+  /// (assumes that the edge type exists)
+  const SetOfEntityTypeIDs& EdgeTypeNameToEntityTypeIDs(
+      const std::string& name) const {
+    return edge_type_name_to_entity_type_ids_.at(name);
+  }
```

Comment:
I think these should be removed. If there is a need to get all supertypes of a given type, there can be a `GetSupertypes` method to do that taking an `EntityTypeID`. That separates name look ups from type relationships.

Reply:
No, they are required. They are used by querying to do set operations like intersection/union/difference on the super-types. 

I like the super-types. I'll change the function name to `NodeTypeNameToSupertypes`

Reply:
Why does the partitioner need to go directly from name to supertypes? We had talked about separating ID look up from supertype look up. So the query can do `GetSupertypes(GetEntityTypeID(name))`.

Reply:
Querying, not partitioner (maybe that's what you meant). 

I agree that the function call `GetSupertypes(GetEntityTypeID(name))` is better.  The datatype returned by `GetSupertypes` needs to be a bitset because querying does set operations like intersection/union/difference on the super-types.

I'll make another pass over it. I understand your requirements better now.

Reply:
I did mean querying. I just did a thinko. :-)

As long as the bitset can be viewed and operated on exactly like a set of EntityTypeIDs. I'm fine with that. A bitset representation makes sense. I only care about the API it provides.

Reply:
Great, then I'll drop the "range" functions.

If I recall, I assumed that you didn't want the representation of a "set" to creep into the API. In this case, it would. That's fine, right?

Reply:
Sure. Smaller API is generally better. So I'm in favor of dropping methods. :-)